inc is intended to increment in-place

o- opened the pull request:
I remembered, that inc_ is supposed to update in-place, even if the
refcount is 1.

That was a feature, not a bug, to be able to increment the loop counter
without creating a new vector all the time.

So the idea is we compile the for loop as follows:

    %i = 0
    while (%i < size(seq))
      ...
      %i++

now, we need to mark the (stack) value %i ensureNamed, because
otherwise the < comparison might override it (values with refcount 0
can be reused). But then the i++ at the end of the loop, should
still be able to increment i without duplicating it. so the inc
bytecode is allowed to destructively modify it

but, the pir version of inc was broken, because the pir optimizer did
not preserve this invariant, that an inc'd value is not used anymore
after the operation.

for now I replaced it by a normal add. this might have potential
performance issues, but we need to solve them differently.

o- committed:
inc is intended to increment in-place

I remembered, that inc_ is supposed to update in-place, even if the
refcount is 1.

That was a feature, not a bug, to be able to increment the loop counter
without creating a new vector all the time.

So the idea is we compile the for loop as follows:

    %i = 0
    while (%i < size(seq))
      ...
      %i++

now, we need to mark the (stack) value %i ensureNamed, because
otherwise the < comparison might override it (values with refcount 0
can be reused). But then the i++ at the end of the loop, should
still be able to increment i without duplicating it. so the inc
bytecode is allowed to destructively modify it

but, the pir version of inc was broken, because the pir optimizer did
not preserve this invariant, that an inc'd value is not used anymore
after the operation.

for now I replaced it by a normal add. this might have potential
performance issues, but we need to solve them differently.

o- commented:
fixes broken #407

o- committed:
re-add inc and fix refcount for subassign

subassign are allowed to override values with refcount 1. This is to
ensure that:

    ldvar i
    subassign ...
    stvar i

does not actually need to copy the vector, even though it has a name
(here "i"). But of course if pir would end up optimizing a subassign and
e.g. do:

    ldvar i
    subassign ...
    dup
    stvar i
    subassign ...
    stvar j

then this would be catastrophal. So, for this case we need to ensure
that the refcount is actually at least 2 for the subassigned array.

The same is true for the inc instruction.

o- commented:
just realized that the same problem applies to subassign, they are also allowed to update inplace. so I had to actually fix it properly at the pir level. whenever an input to a in-place modifying bc is used elsewhere, we need to bump the named count to 2.

Jakobeha commented on rir/src/compiler/analysis/refrence_count.h:
```diff
@@ -98,8 +98,28 @@ class StaticReferenceCount : public StaticAnalysis<AUses> {
+        case Tag::Inc:
+            if (auto input = Instruction::Cast(i->arg(0).val())) {
+                if (state.uses.count(input) && state.uses.at(input) == 2) {
+                    state.uses[input] = 3;
```
maybe rename `uses`, since an instruction will have `uses = 2` if it's used 2+ times but not by a special opcode, and 3+ if it's used 2* by a special opcode.

Also, if `uses = 1` would the special opcode still override? I understand instructions are processed in order, but is this handled:
```r
while (...) {
   use(i)
   i[[n]] <- ...
}
```

o- replied:
the static analysis is a fix-point computation, not one pass. so it will take two passes:

first pass it goes from `0 -> 1 -> 2`, then second pass `2 -> 2 -> 3`

Jakobeha commented on rir/src/ir/insns.h:
```diff
@@ -242,7 +242,7 @@ DEF_INSTR(add_, 0, 2, 1, 0)
 
 /**
- * inc_ :: increment tos integer
+ * inc_ :: increment tos integer (destructive! ignores refcount)
```
Also should mention this for subassign

Jakobeha commented on rir/src/interpreter/interp.cpp:
```diff
@@ -2188,15 +2188,19 @@ SEXP evalRirCode(Code* c, InterpreterInstance* ctx, SEXP env,
+            // Inc_ destructively modifies TOS, even if the refcount is 1. This
+            // can be used to perform `++i` on a value on the stack. The old i
+            // value will be overwritten (generally only do this if you are sure
+            // that this is the last copy on the stack).
```
I don't think this is necessary because you mention the same in `insns.h`. Otherwise, also mention for `subassign`

o- replied:
I think this is *very* neccessary, since we both tried to "fix" this

o- committed:
address review and also remove pir::SetShared

with the new static refcount also setShared can be completely
reconstructed by the pir backend.

mhyee commented:
Typo in the filename: `refrence_count.h` -> `reference_count.h`

o- committed:
fix name and refcount of loop sequence

o- committed:
fix stupidities

arguments which do not need reference count, need not be counted in any
of the cases. Also LdArg and Force always produce shared values in the
first place.

otoh ldvar can return values of named count just 1
